### PR TITLE
Update poh_config to have hashrate be private, with a getter and setter

### PR DESCRIPTION
**Pull request by jdavis103**
In order to increase the hashes-per-tick rate, we need to add a feature gate on it. Since it's a constant, we can't directly add a gate; instead we need to gate the getter for the struct that holds it.

Added getter and setter for hashes-per-tick in poh_config struct. Added "new" as well, to support existing usage by test functions. Updated code and tests.

There should be no functional change to any code.

### CODE REVIEW

File: core/src/banking_stage.rs [hunk 1]
```diff
@@ -2266,10 +2266,9 @@ mod tests {
                 Blockstore::open(ledger_path.path())
                     .expect("Expected to be able to open database ledger"),
             );
-            let poh_config = PohConfig {
-                target_tick_count: Some(bank.max_tick_height() + num_extra_ticks),
-                ..PohConfig::default()
-            };
+            let mut poh_config = PohConfig::default();
```
Comment (taozhu-chicago) [inline core/src/banking_stage.rs:2269]: Maybe overlooking something, but why this change?

Comment (jdavis103) [reply core/src/banking_stage.rs:2269]: When one of the member variables is non-pub, the ".." method of filling in values no longer works. So instead, set it to default, then override the value that you want to be different.

File: sdk/src/poh_config.rs [hunk 2]
```diff
@@ -15,10 +15,30 @@ pub struct PohConfig {
     /// None enables "Low power mode", which implies:
     /// * sleep for `target_tick_duration` instead of hashing
     /// * the number of hashes per tick will be variable
-    pub hashes_per_tick: Option<u64>,
+    hashes_per_tick: Option<u64>,
```
Comment (taozhu-chicago) [inline sdk/src/poh_config.rs:18]: Probably not great to make one field of a simple structure different from others. Maybe make all fields private with getter/setter? But that'd probably touch a lot of code. 

On the other hand, if this PR's goal is to make it possible to default `poh_config.hashes_per_tick` to different values based on a feature gate, would it be possible to leave `poh_config` unchanged, but introduce a function that does something like this
```
fn get_default_hashes_per_tick( feature_set ) -> u64 {
    if feature_set.is_enabled(...) {
        NEW_DEFAULT_HASHES_PER_TICK
    } else {
        clock::DEFAULT_HASHES_PER_TICK
    }
} 
```
Then use `get_default_hashes_per_tick()` wherever constant is used today.

**Comment by KirillLykov** [comment]
Regarding naming, I would stick with https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#getter/setter-methods-[rfc-344] :
```
    A method foo(&self) -> &T for getting the current value of the field.
    A method set_foo(&self, val: T) for setting the field. (The val argument here may take &T or some other type, depending on the context.)

```

File: sdk/src/poh_config.rs (hunk 2 above)
Comment (jdavis103) [reply sdk/src/poh_config.rs:18]: Yeah, interesting idea for sure, but it would be very easy for a future developer to just set any value they want at any point, and if they didn't use get_default_hashes_per_tick, things wouldn't work. The private member is the only way I could see to make this future-programmer-proof.
As for also making the other values private.. yeah, I agree it's a little unbalanced-looking to have one priv and the others pub, but there's nothing really wrong with that, and I didn't want to touch another set of files for a change that didn't do anything we need.

Comment (jdavis103) [reply sdk/src/poh_config.rs:18]: @KirillLykov , I'll make the naming change unless we (@taozhu-chicago) decide that there's a way to make get_default() to work sensibly. Because if so, that would be a much smaller code change.

Comment (KirillLykov) [reply sdk/src/poh_config.rs:18]: Yeah, this is more about style and so on, so the best approach to do similar to the other parts of the code, I guess.

**Comment by jdavis103** [comment]
> Regarding naming, I would stick with [https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#getter/setter-methods-[rfc-344]](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#getter/setter-methods-%5Brfc-344%5D) :
> 
> ```
>     A method foo(&self) -> &T for getting the current value of the field.
>     A method set_foo(&self, val: T) for setting the field. (The val argument here may take &T or some other type, depending on the context.)
> ```

Looking through the codebase, "get_" is used quite a bit, so I think I'm okay with leaving that alone.

**Comment by t-nelson** [comment]
i don't think this is the right way forward.  we need to remove all logic that requires _holding onto_ a `PohConfig`.  `PohConfig` should be create, instantiate poh recorder/service, drop.  that it's wrapped in an `Arc` is a bad sign.  we're not going to be able to reinitialize poh at runtime without this as a first step (or a bunch of gross synchronization).

generally `*Config` should have all members pub to ease instantiating various configuration

**Comment by t-nelson** [comment]
> i don't think this is the right way forward. we need to remove all logic that requires _holding onto_ a `PohConfig`. `PohConfig` should be create, instantiate poh recorder/service, drop. that it's wrapped in an `Arc` is a bad sign. we're not going to be able to reinitialize poh at runtime without this as a first step (or a bunch of gross synchronization).
> 
> generally `*Config` should have all members pub to ease instantiating various configuration

looks like we're really most of the way there already. no one who's holding an instance today needs to, aside from `GenesisConfig`, which _should_ have similar `*Config` semantics as described above

File: sdk/src/poh_config.rs [hunk 3]
```diff
@@ -3,7 +3,7 @@ use {
     std::time::Duration,
 };
 
-#[derive(Serialize, Deserialize, Clone, Debug, AbiExample, Eq, PartialEq)]
+#[derive(Serialize, Deserialize, Clone, Copy, Debug, AbiExample, Eq, PartialEq)]
```
Comment (t-nelson) [inline sdk/src/poh_config.rs:6]: `PohConfig` is a little large for `Copy`.  Prefer passing them around by reference and `clone()`ing where we need to keep one.

File: sdk/src/poh_config.rs (hunk 2 above)
Comment (t-nelson) [inline sdk/src/poh_config.rs:18]: let's revert this member privacy, the getter/setter/new and all related changes.  it breaks the `*Config` ergonomics that are expected elsewhere in the code base

**Review by t-nelson — COMMENTED** [review]
can you try to rebase against master instead of merging?  looks like some bad conflict resolution picked up unrelated changes

**Comment by jdavis103** [comment]
> can you try to rebase against master instead of merging? looks like some bad conflict resolution picked up unrelated changes

Hmm, I did do a rebase to pick up the latest master (though I realize that my comment says "merge" instead).

File: sdk/src/poh_config.rs (hunk 2 above)
Comment (t-nelson) [reply sdk/src/poh_config.rs:18]: i was thinking we just reinitialize the service with a new `PohConfig`

**Comment by t-nelson** [comment]
> Hmm, I did do a rebase to pick up the latest master (though I realize that my comment says "merge" instead).

weird.  usually don't get unrelated changes from a rebase that aren't directly adjacent to intended changes.

either way, seeing those unrelated changes and all of the plumbing noise from the superfluous private member changes makes it difficult for me to be confident in my review

**Comment by jdavis103** [comment]
> either way, seeing those unrelated changes and all of the plumbing noise from the superfluous private member changes makes it difficult for me to be confident in my review

I'll make your suggested adjustments and redo this as a new PR so it's easier to follow.

**Comment by t-nelson** [comment]
new PR isn't necessary.  you can just force push to this branch and it'll overwrite

**Comment by jdavis103** [comment]
> new PR isn't necessary. you can just force push to this branch and it'll overwrite

Was thinking that with the weird rebase thing of pulling in something as a change that wasn't, it might make sense to just clear it.
